File: backend/app/schemas/weather.py
"""Weather schemas for API validation"""
from pydantic import BaseModel, Field
from typing import Optional
from datetime import datetime


class WeatherData(BaseModel):
    """Weather data model for a jump location and time"""
    temperature_celsius: Optional[float] = Field(None, description="Temperature in Celsius")
    wind_speed_kmh: Optional[float] = Field(None, description="Wind speed in km/h")
    wind_direction_degrees: Optional[int] = Field(None, ge=0, lt=360, description="Wind direction in degrees (0-359)")
    wind_gusts_kmh: Optional[float] = Field(None, description="Wind gusts in km/h")
    weather_code: Optional[int] = Field(None, description="WMO weather code")
    weather_description: Optional[str] = Field(None, description="Human readable weather description")
    humidity_percent: Optional[int] = Field(None, ge=0, le=100, description="Relative humidity in percent")
    pressure_hpa: Optional[float] = Field(None, description="Surface pressure in hPa")
    cloud_cover_percent: Optional[int] = Field(None, ge=0, le=100, description="Cloud cover in percent")
    visibility_km: Optional[float] = Field(None, description="Visibility in kilometers")

    @staticmethod
    def get_wind_direction_name(degrees: Optional[int]) -> Optional[str]:
        """Convert wind direction degrees to compass direction"""
        if degrees is None:
            return None
        directions = ["N", "NNE", "NE", "ENE", "E", "ESE", "SE", "SSE",
                      "S", "SSW", "SW", "WSW", "W", "WNW", "NW", "NNW"]
        index = round(degrees / 22.5) % 16
        return directions[index]
    
    @staticmethod
    def get_weather_description_from_code(code: Optional[int]) -> Optional[str]:
        """Convert WMO weather code to human readable description"""
        if code is None:
            return None
        
        weather_codes = {
            0: "Klar",
            1: "Überwiegend klar",
            2: "Teilweise bewölkt",
            3: "Bewölkt",
            45: "Nebel",
            48: "Gefrierender Nebel",
            51: "Leichter Nieselregen",
            53: "Mäßiger Nieselregen",
            55: "Starker Nieselregen",
            56: "Leichter gefrierender Nieselregen",
            57: "Starker gefrierender Nieselregen",
            61: "Leichter Regen",
            63: "Mäßiger Regen",
            65: "Starker Regen",
            66: "Leichter gefrierender Regen",
            67: "Starker gefrierender Regen",
            71: "Leichter Schneefall",
            73: "Mäßiger Schneefall",
            75: "Starker Schneefall",
            77: "Schneekörner",
            80: "Leichte Regenschauer",
            81: "Mäßige Regenschauer",
            82: "Heftige Regenschauer",
            85: "Leichte Schneeschauer",
            86: "Starke Schneeschauer",
            95: "Gewitter",
            96: "Gewitter mit leichtem Hagel",
            99: "Gewitter mit starkem Hagel",
        }
        return weather_codes.get(code, f"Unbekannt ({code})")


class WeatherRequest(BaseModel):
    """Request model for fetching weather data"""
    latitude: float = Field(..., ge=-90, le=90, description="Latitude")
    longitude: float = Field(..., ge=-180, le=180, description="Longitude")
    datetime: datetime = Field(..., description="Date and time for weather data")


class WeatherResponse(BaseModel):
    """Response model for weather data"""
    weather: Optional[WeatherData] = None
    success: bool = True
    error: Optional[str] = None

File: frontend/lib/models/weather.dart
/// Weather data model for jump weather conditions
class WeatherData {
  final double? temperatureCelsius;
  final double? windSpeedKmh;
  final int? windDirectionDegrees;
  final double? windGustsKmh;
  final int? weatherCode;
  final String? weatherDescription;
  final int? humidityPercent;
  final double? pressureHpa;
  final int? cloudCoverPercent;
  final double? visibilityKm;

  WeatherData({
    this.temperatureCelsius,
    this.windSpeedKmh,
    this.windDirectionDegrees,
    this.windGustsKmh,
    this.weatherCode,
    this.weatherDescription,
    this.humidityPercent,
    this.pressureHpa,
    this.cloudCoverPercent,
    this.visibilityKm,
  });

  /// Get wind direction as compass direction (N, NE, E, etc.)
  String? get windDirectionName {
    if (windDirectionDegrees == null) return null;
    const directions = [
      'N', 'NNO', 'NO', 'ONO', 'O', 'OSO', 'SO', 'SSO',
      'S', 'SSW', 'SW', 'WSW', 'W', 'WNW', 'NW', 'NNW'
    ];
    final index = ((windDirectionDegrees! / 22.5).round()) % 16;
    return directions[index];
  }

  /// Check if we have any weather data
  bool get hasData =>
      temperatureCelsius != null ||
      windSpeedKmh != null ||
      windDirectionDegrees != null ||
      weatherCode != null;

  /// Check if we have wind data
  bool get hasWindData =>
      windSpeedKmh != null || windDirectionDegrees != null;

  Map<String, dynamic> toMap() {
    return {
      'temperature_celsius': temperatureCelsius,
      'wind_speed_kmh': windSpeedKmh,
      'wind_direction_degrees': windDirectionDegrees,
      'wind_gusts_kmh': windGustsKmh,
      'weather_code': weatherCode,
      'weather_description': weatherDescription,
      'humidity_percent': humidityPercent,
      'pressure_hpa': pressureHpa,
      'cloud_cover_percent': cloudCoverPercent,
      'visibility_km': visibilityKm,
    };
  }

  factory WeatherData.fromMap(Map<String, dynamic>? map) {
    if (map == null || map.isEmpty) return WeatherData();

    return WeatherData(
      temperatureCelsius: _parseDouble(map['temperature_celsius']),
      windSpeedKmh: _parseDouble(map['wind_speed_kmh']),
      windDirectionDegrees: _parseInt(map['wind_direction_degrees']),
      windGustsKmh: _parseDouble(map['wind_gusts_kmh']),
      weatherCode: _parseInt(map['weather_code']),
      weatherDescription: map['weather_description'] as String?,
      humidityPercent: _parseInt(map['humidity_percent']),
      pressureHpa: _parseDouble(map['pressure_hpa']),
      cloudCoverPercent: _parseInt(map['cloud_cover_percent']),
      visibilityKm: _parseDouble(map['visibility_km']),
    );
  }

  static double? _parseDouble(dynamic value) {
    if (value == null) return null;
    if (value is double) return value;
    if (value is int) return value.toDouble();
    if (value is String) {
      try {
        return double.parse(value);
      } catch (e) {
        return null;
      }
    }
    return null;
  }

  static int? _parseInt(dynamic value) {
    if (value == null) return null;
    if (value is int) return value;
    if (value is double) return value.toInt();
    if (value is String) {
      try {
        return int.parse(value);
      } catch (e) {
        return null;
      }
    }
    return null;
  }

  WeatherData copyWith({
    double? temperatureCelsius,
    double? windSpeedKmh,
    int? windDirectionDegrees,
    double? windGustsKmh,
    int? weatherCode,
    String? weatherDescription,
    int? humidityPercent,
    double? pressureHpa,
    int? cloudCoverPercent,
    double? visibilityKm,
  }) {
    return WeatherData(
      temperatureCelsius: temperatureCelsius ?? this.temperatureCelsius,
      windSpeedKmh: windSpeedKmh ?? this.windSpeedKmh,
      windDirectionDegrees: windDirectionDegrees ?? this.windDirectionDegrees,
      windGustsKmh: windGustsKmh ?? this.windGustsKmh,
      weatherCode: weatherCode ?? this.weatherCode,
      weatherDescription: weatherDescription ?? this.weatherDescription,
      humidityPercent: humidityPercent ?? this.humidityPercent,
      pressureHpa: pressureHpa ?? this.pressureHpa,
      cloudCoverPercent: cloudCoverPercent ?? this.cloudCoverPercent,
      visibilityKm: visibilityKm ?? this.visibilityKm,
    );
  }
}

